src/ray/util/process.cc

```diff
@@ -63,6 +68,123 @@ int execvpe(const char *program, char *const argv[], char *const envp[]) {
 
 namespace ray {
 
+namespace {
+// Global tracker of owned child processes.
+// Only tracked in linux.
+//
+// A pid is added when a spawn is called and removed when the ProcessFD is closed (i.e.
+// it's in dtor and we no longer want to track it).
+std::mutex m;
+absl::flat_hash_set<pid_t> owned_children;
+
+void addOwnedChild(pid_t pid) {
+#ifdef __linux__
+  std::lock_guard<std::mutex> guard(m);
+  owned_children.insert(pid);
+#endif
+}
+
+void removeOwnedChild(pid_t pid) {
+#ifdef __linux__
+  std::lock_guard<std::mutex> guard(m);
+  owned_children.erase(pid);
+#endif
+}
+
+// Register a signal handler for the given signal.
+// The handler will be called with the signal_set and the error code.
+// After the handler is called, the signal will be re-registered.
+void RegisterSignalHandlerLoop(boost::asio::signal_set &signals,
+                               void (*handler)(const boost::system::error_code &, int)) {
+  signals.async_wait(
+      [&signals, handler](const boost::system::error_code &error, int signal_number) {
+        handler(error, signal_number);
+        RegisterSignalHandlerLoop(signals, handler);
+      });
+}
+
+void SigchldHandlerPlain(const boost::system::error_code &error, int signal_number) {
+  if (!error) {
+    int status;
+    pid_t pid;
+    while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
+      if (WIFEXITED(status)) {
+        RAY_LOG(INFO) << "Child process " << pid << " exited with status "
+                      << WEXITSTATUS(status);
+      } else if (WIFSIGNALED(status)) {
+        RAY_LOG(INFO) << "Child process " << pid << " exited from signal "
+                      << WTERMSIG(status);
+      }
+      removeOwnedChild(pid);
+    }
+  }
+}
+
+#ifdef __linux__
+
+void KillUnownedChildren() {
+  auto maybe_child_procs = GetAllProcsWithPpid(GetPID());
+
+  // Enumerating child procs is not supported on this platform.
+  if (!maybe_child_procs) {
+    RAY_LOG(WARNING)
+        << "Killing leaked procs not supported on this platform. Only supports Linux";
+    return;
+  }
+
+  std::vector<pid_t> to_kill;
+  to_kill.reserve(maybe_child_procs->size());
+  {
+    std::lock_guard<std::mutex> guard(m);
+    for (auto pid : *maybe_child_procs) {
+      if (owned_children.count(pid) == 0) {
+        to_kill.push_back(pid);
+      }
+    }
+  }
+  for (auto pid : to_kill) {
+    if (owned_children.count(pid) == 0) {
+      RAY_LOG(INFO) << "Killing leaked child process " << pid;
+      auto error = KillProc(pid);
+      if (error) {
+        RAY_LOG(ERROR) << "Failed to kill leaked child process " << pid << " with error "
+                       << error->message();
+      }
+    }
+  }
+}
+
+// Set this process as a subreaper.
+void SetThisProcessAsSubreaper() {
+  if (prctl(PR_SET_CHILD_SUBREAPER, 1) == -1) {
```

Comment:
it is only available after linux 3.4. Do you know how old is this version? If it is not supported, does it mean it will just crash raylet?

Reply:
Linux 3.4 is released in 2012. As a comparison: ubuntu 18.04 is on linux 5.4. So we should be safe.